Added SPL token vault tutorial

Woody4618 opened the pull request:
- Learn how to create SPL tokens with meta data
- Save tokens in an anchor program using PDAs

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/client.ts.raw:
```diff
@@ -0,0 +1,98 @@
+
+const decimals = 9;
+
+let [tokenAccountOwnerPda, bump] = await PublicKey.findProgramAddress(
```
Can we use the sync version of this(deprecated)?
```suggestion
let [tokenAccountOwnerPda] =  PublicKey.findProgramAddressSync(
```

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/client.ts.raw:
```diff
@@ -0,0 +1,98 @@
+const decimals = 9;
+
+let [tokenAccountOwnerPda, bump] = await PublicKey.findProgramAddress(
+  [Buffer.from("token_account_owner_pda", "utf8")],
```
```suggestion
  [Buffer.from("token_account_owner_pda")],
```

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/client.ts.raw:
```diff
@@ -0,0 +1,98 @@
+console.log(
+  "Owned token amount: " + tokenAccountInfo.amount / BigInt(mintDecimals)
+);
+let [token_vault, bump2] = await PublicKey.findProgramAddress(
```
Same here, also better to keep variable names camelCase for consistency.
```suggestion
let [tokenVault] = PublicKey.findProgramAddressSync(
```

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/lib.rs:
```diff
@@ -0,0 +1,86 @@
+        Ok(())
+    }
+
+    pub fn transfer_out(ctx: Context<TransferAccounts>, amount: u64) -> Result<()> {
```
Warning.

```suggestion
    pub fn transfer_out(ctx: Context<TransferAccounts>, amount: u64) -> Result<()> {
```

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/lib.rs:
```diff
@@ -0,0 +1,86 @@
+    #[account(
+            init_if_needed,
+            payer = signer,
+            seeds=[b"token_vault".as_ref(), mint_of_token_being_sent.key().as_ref()],
```
```suggestion
            seeds=[b"token_vault", mint_of_token_being_sent.key().as_ref()],
```

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/lib.rs:
```diff
@@ -0,0 +1,86 @@
+pub struct TransferAccounts<'info> {
+    // Derived PDAs
+    #[account(mut,
+            seeds=[b"token_account_owner_pda".as_ref()],
```
```suggestion
            seeds=[b"token_account_owner_pda"],
```

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/lib.rs:
```diff
@@ -0,0 +1,86 @@
+    token_account_owner_pda: AccountInfo<'info>,
+
+    #[account(mut,
+            seeds=[b"token_vault".as_ref(), mint_of_token_being_sent.key().as_ref()],
```
```suggestion
            seeds=[b"token_vault", mint_of_token_being_sent.key().as_ref()],
```

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/lib.rs:
```diff
@@ -0,0 +1,86 @@
+#[derive(Accounts)]
+pub struct Initialize<'info> {
+    // Derived PDAs
+    #[account(
```
Inside is extra indented.

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/lib.rs:
```diff
@@ -0,0 +1,86 @@
+        )]
+    token_account_owner_pda: AccountInfo<'info>,
+
+    #[account(
```
Inside is extra indented.

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/lib.rs:
```diff
@@ -0,0 +1,86 @@
+#[derive(Accounts)]
+pub struct TransferAccounts<'info> {
+    // Derived PDAs
+    #[account(mut,
```
Inside is extra indented.

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/lib.rs:
```diff
@@ -0,0 +1,86 @@
+        )]
+    token_account_owner_pda: AccountInfo<'info>,
+
+    #[account(mut,
```
Inside is extra indented.

Woody4618 commented on client/src/tutorials/SplTokenVault/files/lib.rs:
```diff
@@ -0,0 +1,86 @@
+use anchor_lang::prelude::*;
+use anchor_spl::token::{Mint, Token, TokenAccount, Transfer};
```
These warnings come from the code that will be pasted in during the tutorial. Maybe it would be better to have all the code in the files already? It maybe confusing to people if the have to add imports and change _ctx to ctx etc. 
So either leave the warning or already put the whole code it. I tend to having all the code in because it makes it easier to work with and an easy reference when you want to look things up.

Woody4618 replied:
I removed the warning and added to the tutorial text to add transfer to the imports.

acheroncrypto replied:
Ah sorry, I thought it was just a forgotten import.

Woody4618 commented:
All done 👍

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/client.ts.raw:
```diff
@@ -0,0 +1,110 @@
+
+const decimals = 9;
+
+let [tokenAccountOwnerPda] = await PublicKey.findProgramAddressSync(
```
Using sync version.

```suggestion
let [tokenAccountOwnerPda] = PublicKey.findProgramAddressSync(
```

acheroncrypto commented on client/src/tutorials/SplTokenVault/files/client.ts.raw:
```diff
@@ -0,0 +1,110 @@
+console.log(
+  "Owned token amount: " + tokenAccountInfo.amount / BigInt(mintDecimals)
+);
+let [tokenVault] = await PublicKey.findProgramAddressSync(
```
```suggestion
let [tokenVault] = PublicKey.findProgramAddressSync(
```

acheroncrypto reviewed (commented):
In review, I'm only picking the ones in the files but the markdown part also needs to be updated after making changes to program files or tests. This is also why the *Hello World* tutorials only show the code in markdown files so that it doesn't need to be synced.

There is also another advantage of using the tutorial files and let the users write it for themselves which is if a user started a tutorial, the markdown files can be updated anytime and the user will see the changes but the actual program files will not get updated since its just a file that is stored in the user's browser and they are free to do anything they want with it.

I'll think of ways to mitigate this problem, maybe a button to show the completed version of the tutorial works better?

Woody4618 commented:
Updated

Woody4618 commented:
I also updated this example to the newest version. Did I do that correct with that assets submodule? I just added the icon there, i hope that is enough?

acheroncrypto reviewed (commented):
> I also updated this example to the newest version. Did I do that correct with that assets submodule? I just added the icon there, i hope that is enough?

Thanks, please see https://github.com/solana-playground/assets/pull/1#pullrequestreview-1813516898. The PR part is correct but I think you've also updated the submodule to your fork in this PR which shouldn't happen. Instead, run:

```sh
git submodule update --remote
```

**after** [the asset PR](https://github.com/solana-playground/assets/pull/1) is merged.

acheroncrypto reviewed (changes requested):
Thanks, [the asset PR](https://github.com/solana-playground/assets/pull/1) is merged, only need to run:

```sh
git submodule update --remote
```

There is also a discrepancy between the actual code files and the Markdown explanation files. We need to make the changes from both places currently.

Woody4618 commented:
Updated code and public submodule using git submodule update --remote

acheroncrypto reviewed (commented):
> Updated code and public submodule using git submodule update --remote

Thanks, it's correct now.

However, there is still a mismatch between the code files and the Markdown files. I only point them in one place in my reviews but its counterpart should also be updated. I can comment them here if you'd like but it would be much easier if you compared it in your editor.

Woody4618 commented:
I think i changed all the differences between md files and code now. Sorry for the hassle

acheroncrypto reviewed (approved):
Thank you, this will most likely go live at the end of the month.
